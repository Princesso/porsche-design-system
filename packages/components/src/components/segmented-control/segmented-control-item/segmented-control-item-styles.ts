import { getCss } from '../../../utils';
import {
  addImportantToEachRule,
  getInsetJssStyle,
  getThemedColors,
  getTransition,
  pxToRemWithUnit,
} from '../../../styles';
import {
  borderRadiusSmall,
  borderWidthBase,
  textSmallStyle,
  textXSmallStyle,
} from '@porsche-design-system/utilities-v2';
import type { Theme } from '../../../types';
import type { SegmentedControlBackgroundColor } from '../segmented-control/segmented-control-utils';
import { hoverMediaQuery } from '../../../styles/hover-media-query';

export const ITEM_PADDING = pxToRemWithUnit(16);
export const { font: BUTTON_FONT } = textSmallStyle;
export const { font: LABEL_FONT } = textXSmallStyle;
export const ICON_SIZE = pxToRemWithUnit(24);
export const ICON_MARGIN = pxToRemWithUnit(4);

export const getColors = (
  isDisabled: boolean,
  isSelected: boolean,
  bgColor: SegmentedControlBackgroundColor,
  theme: Theme
): {
  backgroundColor: string;
  buttonColor: string;
  labelColor: string;
  borderColor: string;
  hoverBorderColor: string;
} => {
  const themedColors = getThemedColors(theme);

  const { primaryColor, contrastMediumColor } = themedColors;
  const backgroundColor = themedColors[bgColor === 'background-surface' ? 'backgroundColor' : 'backgroundSurfaceColor'];
  const borderColor = isSelected ? themedColors.primaryColor : 'transparent';
  const hoverBorderColor = isSelected ? themedColors.primaryColor : themedColors.contrastMediumColor;

  return isDisabled
    ? {
        backgroundColor,
        buttonColor: themedColors.disabledColor,
        labelColor: themedColors.disabledColor,
        borderColor,
        hoverBorderColor,
      }
    : {
        backgroundColor,
        buttonColor: primaryColor,
        labelColor: contrastMediumColor,
        borderColor,
        hoverBorderColor,
      };
};

export const getComponentCss = (
  isDisabled: boolean,
  isSelected: boolean,
  bgColor: SegmentedControlBackgroundColor,
  theme: Theme
): string => {
  const { contrastLowColor, focusColor } = getThemedColors(theme);
  const { backgroundColor, buttonColor, labelColor, borderColor, hoverBorderColor } = getColors(
    isDisabled,
    isSelected,
    bgColor,
    theme
  );

  return getCss({
    '@global': {
      ':host': addImportantToEachRule({
        display: 'block',
        outline: 0,
      }),
      button: {
        display: 'block',
        height: '100%',
        width: '100%',
        padding: `${pxToRemWithUnit(12)} ${ITEM_PADDING}`,
        margin: 0,
        border: `${borderWidthBase} solid ${borderColor}`,
        transition: getTransition('border-color'),
        outline: 0,
        background: backgroundColor,
        color: buttonColor,
        ...textSmallStyle,
        overflowWrap: 'normal',
        position: 'relative',
        borderRadius: borderRadiusSmall,
        '&::before': {
          content: '""',
          position: 'absolute',
          ...getInsetJssStyle(-5),
          border: `${borderWidthBase} solid transparent`,
          borderRadius: '7px',
        },
        '&:focus::before': {
          borderColor: focusColor,
        },
        '&:focus:not(:focus-visible)::before': {
          borderColor: 'transparent',
        },
        ...(isDisabled
          ? {
              cursor: 'not-allowed',
            }
          : {
              cursor: 'pointer',
              ...hoverMediaQuery({
                transition: getTransition('background-color'),
                '&:hover': {
                  background: contrastLowColor,
                  borderColor: hoverBorderColor,
                },
              }),
            }),
      },
      // label
      span: {
        display: 'block',
        ...textXSmallStyle,
        overflowWrap: 'normal',
        color: labelColor,
      },
    },
    icon: {
      height: ICON_SIZE,
      width: ICON_SIZE,
      marginRight: ICON_MARGIN,
    },
  });
};
