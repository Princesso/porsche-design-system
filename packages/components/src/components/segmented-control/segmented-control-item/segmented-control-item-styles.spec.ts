import { getComponentCss } from './segmented-control-item-styles';

describe('getComponentCss()', () => {
  it.each<Parameters<typeof getComponentCss>>([
    [false, false, 'background-default', true, 'light'],
    [true, false, 'background-default', true, 'light'],
    [true, true, 'background-default', true, 'light'],
    [false, true, 'background-default', true, 'light'],
    [false, false, 'background-surface', false, 'light'],
    [false, false, 'background-default', false, 'dark'],
    [false, false, 'background-surface', false, 'dark'],
  ])(
    'should return correct css for isDisabled: %s, isSelected: %s, bgColor: %s, icon: %s, iconSource: %s and theme: %s',
    (...args) => {
      expect(getComponentCss(...args)).toMatchSnapshot();
    }
  );
});
