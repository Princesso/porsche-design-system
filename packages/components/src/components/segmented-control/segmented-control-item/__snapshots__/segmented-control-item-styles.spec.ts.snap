// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponentCss() should return correct css for isDisabled: false, isSelected: false, bgColor: background-default and theme: dark 1`] = `
":host {
  display: block !important;
  outline: 0 !important;
}
button {
  display: block;
  height: 100%;
  width: 100%;
  padding: 13px 17px;
  margin: 0;
  border: 2px solid #7E7F82;
  transition: border-color var(--p-transition-duration, .24s) ease;
  outline: 0;
  background-color: transparent;
  color: #FBFCFF;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  position: relative;
  border-radius: 4px;
  cursor: pointer;
}
button::before {
  content: "";
  position: absolute;
  top: -5px;
  left: -5px;
  right: -5px;
  bottom: -5px;
  border: 2px solid transparent;
  border-radius: 7px;
}
button:focus::before {
  border-color: #0A0AFF;
}
button:focus:not(:focus-visible)::before {
  border-color: transparent;
}
span {
  display: block;
  font: normal normal 400 .875rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  color: #7E7F82;
}
@media(hover:hover) {
  button:hover {
    border-color: #FBFCFF;
    cursor: pointer;
  }
}

.icon {
  height: 1.5rem;
  width: 1.5rem;
  margin-right: 0.25rem;
}"
`;

exports[`getComponentCss() should return correct css for isDisabled: false, isSelected: false, bgColor: background-default and theme: light 1`] = `
":host {
  display: block !important;
  outline: 0 !important;
}
button {
  display: block;
  height: 100%;
  width: 100%;
  padding: 13px 17px;
  margin: 0;
  border: 2px solid #949598;
  transition: border-color var(--p-transition-duration, .24s) ease;
  outline: 0;
  background-color: transparent;
  color: #010205;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  position: relative;
  border-radius: 4px;
  cursor: pointer;
}
button::before {
  content: "";
  position: absolute;
  top: -5px;
  left: -5px;
  right: -5px;
  bottom: -5px;
  border: 2px solid transparent;
  border-radius: 7px;
}
button:focus::before {
  border-color: #0A0AFF;
}
button:focus:not(:focus-visible)::before {
  border-color: transparent;
}
span {
  display: block;
  font: normal normal 400 .875rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  color: #949598;
}
@media(hover:hover) {
  button:hover {
    border-color: #010205;
    cursor: pointer;
  }
}

.icon {
  height: 1.5rem;
  width: 1.5rem;
  margin-right: 0.25rem;
}"
`;

exports[`getComponentCss() should return correct css for isDisabled: false, isSelected: false, bgColor: background-surface and theme: dark 1`] = `
":host {
  display: block !important;
  outline: 0 !important;
}
button {
  display: block;
  height: 100%;
  width: 100%;
  padding: 13px 17px;
  margin: 0;
  border: 2px solid #7E7F82;
  transition: border-color var(--p-transition-duration, .24s) ease;
  outline: 0;
  background-color: transparent;
  color: #FBFCFF;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  position: relative;
  border-radius: 4px;
  cursor: pointer;
}
button::before {
  content: "";
  position: absolute;
  top: -5px;
  left: -5px;
  right: -5px;
  bottom: -5px;
  border: 2px solid transparent;
  border-radius: 7px;
}
button:focus::before {
  border-color: #0A0AFF;
}
button:focus:not(:focus-visible)::before {
  border-color: transparent;
}
span {
  display: block;
  font: normal normal 400 .875rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  color: #7E7F82;
}
@media(hover:hover) {
  button:hover {
    border-color: #FBFCFF;
    cursor: pointer;
  }
}

.icon {
  height: 1.5rem;
  width: 1.5rem;
  margin-right: 0.25rem;
}"
`;

exports[`getComponentCss() should return correct css for isDisabled: false, isSelected: false, bgColor: background-surface and theme: light 1`] = `
":host {
  display: block !important;
  outline: 0 !important;
}
button {
  display: block;
  height: 100%;
  width: 100%;
  padding: 13px 17px;
  margin: 0;
  border: 2px solid #949598;
  transition: border-color var(--p-transition-duration, .24s) ease;
  outline: 0;
  background-color: transparent;
  color: #010205;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  position: relative;
  border-radius: 4px;
  cursor: pointer;
}
button::before {
  content: "";
  position: absolute;
  top: -5px;
  left: -5px;
  right: -5px;
  bottom: -5px;
  border: 2px solid transparent;
  border-radius: 7px;
}
button:focus::before {
  border-color: #0A0AFF;
}
button:focus:not(:focus-visible)::before {
  border-color: transparent;
}
span {
  display: block;
  font: normal normal 400 .875rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  color: #949598;
}
@media(hover:hover) {
  button:hover {
    border-color: #010205;
    cursor: pointer;
  }
}

.icon {
  height: 1.5rem;
  width: 1.5rem;
  margin-right: 0.25rem;
}"
`;

exports[`getComponentCss() should return correct css for isDisabled: false, isSelected: true, bgColor: background-default and theme: light 1`] = `
":host {
  display: block !important;
  outline: 0 !important;
}
button {
  display: block;
  height: 100%;
  width: 100%;
  padding: 13px 17px;
  margin: 0;
  border: 2px solid #010205;
  transition: border-color var(--p-transition-duration, .24s) ease;
  outline: 0;
  background-color: transparent;
  color: #010205;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  position: relative;
  border-radius: 4px;
  cursor: pointer;
}
button::before {
  content: "";
  position: absolute;
  top: -5px;
  left: -5px;
  right: -5px;
  bottom: -5px;
  border: 2px solid transparent;
  border-radius: 7px;
}
button:focus::before {
  border-color: #0A0AFF;
}
button:focus:not(:focus-visible)::before {
  border-color: transparent;
}
span {
  display: block;
  font: normal normal 400 .875rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  color: #949598;
}
@media(hover:hover) {
  button:hover {
    border-color: #010205;
    cursor: pointer;
  }
}

.icon {
  height: 1.5rem;
  width: 1.5rem;
  margin-right: 0.25rem;
}"
`;

exports[`getComponentCss() should return correct css for isDisabled: true, isSelected: false, bgColor: background-default and theme: light 1`] = `
":host {
  display: block !important;
  outline: 0 !important;
}
button {
  display: block;
  height: 100%;
  width: 100%;
  padding: 13px 17px;
  margin: 0;
  border: 2px solid #949598;
  transition: border-color var(--p-transition-duration, .24s) ease;
  outline: 0;
  background-color: transparent;
  color: #949598;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  position: relative;
  border-radius: 4px;
  cursor: not-allowed;
}
button::before {
  content: "";
  position: absolute;
  top: -5px;
  left: -5px;
  right: -5px;
  bottom: -5px;
  border: 2px solid transparent;
  border-radius: 7px;
}
button:focus::before {
  border-color: #0A0AFF;
}
button:focus:not(:focus-visible)::before {
  border-color: transparent;
}
span {
  display: block;
  font: normal normal 400 .875rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  color: #949598;
}
.icon {
  height: 1.5rem;
  width: 1.5rem;
  margin-right: 0.25rem;
}"
`;

exports[`getComponentCss() should return correct css for isDisabled: true, isSelected: true, bgColor: background-default and theme: light 1`] = `
":host {
  display: block !important;
  outline: 0 !important;
}
button {
  display: block;
  height: 100%;
  width: 100%;
  padding: 13px 17px;
  margin: 0;
  border: 2px solid #010205;
  transition: border-color var(--p-transition-duration, .24s) ease;
  outline: 0;
  background-color: transparent;
  color: #949598;
  font: normal normal 400 1rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  position: relative;
  border-radius: 4px;
  cursor: not-allowed;
}
button::before {
  content: "";
  position: absolute;
  top: -5px;
  left: -5px;
  right: -5px;
  bottom: -5px;
  border: 2px solid transparent;
  border-radius: 7px;
}
button:focus::before {
  border-color: #0A0AFF;
}
button:focus:not(:focus-visible)::before {
  border-color: transparent;
}
span {
  display: block;
  font: normal normal 400 .875rem/calc(6px + 2.125ex) 'Porsche Next','Arial Narrow',Arial,'Heiti SC',SimHei,sans-serif;
  overflow-wrap: normal;
  hyphens: auto;
  color: #949598;
}
.icon {
  height: 1.5rem;
  width: 1.5rem;
  margin-right: 0.25rem;
}"
`;
